563 - Pull container images

alexandre-allard opened the pull request.

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @thomasdanan

* @ramanduh

* @zempashi

* @alexandre-allard-scality

* @slaperche-scality

* @sayf-eddine-scality

* @alxf

* @gdemonet

* @tcarmet

* @monpote

* @ballot-scality

* @teddyandrieux

* @cmonfort

* @nicolast

* @nootal

* @carlito-scality

* @lucieleonard

alexandre-allard commented on Makefile:
```diff
@@ -17,6 +17,15 @@ ISO ?= $(BUILD_ROOT)/metalk8s.iso
+	k8s.gcr.io/kube-proxy@sha256:8886e69a7946717a88cad9882e9b44927798d154088be19b5116a375c9923036 \
+	k8s.gcr.io/etcd@sha256:b960569ade5f37205a033dcdc3191fe99dc95b15c6795a6282859070ec2c6124 \
+	k8s.gcr.io/coredns@sha256:3e2be1cec87aca0b74b7668bbe8c02964a95a402e45ceb51b2252629d608d03a \
+	nginx@sha256:dd2d0ac3fff2f007d99e033b64854be0941e19a2ad51f174d9240dda20d9f534
```
Not sure about the image of Nginx we should pull, this is the nginx:latest, maybe we want another one.

ballot-scality replied:
The Makefile start to be a bit huge.

I would prefer to split the files based on their purpose.
a "container_images.mk" with the list.
Ideally, the file name are the same as the images (if that makes sense), so you can really trigger the right download for the given image if we change it

alexandre-allard replied:
done

slaperche-scality commented on scripts/get_container_images.sh:
```diff
@@ -0,0 +1,70 @@
+declare -r LONG_OPTS='destination:'
+declare -r SHORT_OPTS='d:'
+
+print_error() {
```
Shouldn't error be printed on stderr?

alexandre-allard replied:
Right ! :)

ballot-scality replied:
I second this

ballot-scality commented on scripts/get_container_images.sh:
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+
```
set -e ?

alexandre-allard replied:
Not really useful since I already check all the return codes of the executed commands, but I can add a "set -eu" anyway.

ballot-scality commented on Makefile:
```diff
@@ -232,3 +243,7 @@ $(SCALITY_EL7_REPODATA): $(SCALITY_EL7_RPMS) | $(PACKAGE_BUILD_CONTAINER)
 		$(PACKAGE_BUILD_IMAGE) \
 		/entrypoint.sh buildrepo
+
+$(ISO_ROOT)/images:
```
I would put as dependency for the target all the files here
Right now the target won't be executed a second time if the directory exists and isn't changed.

alexandre-allard replied:
done

ballot-scality reviewed (changes requested):
Small changes required for the makefile

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @ramanduh

* @alxf

* @tcarmet

* @slaperche-scality

* @cmonfort

* @zempashi

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @alexandre-allard-scality

* @ballot-scality

* @sayf-eddine-scality

* @lucieleonard

* @gdemonet

* @carlito-scality

* @teddyandrieux





The following reviewers are expecting changes from the author, or must review again:

* @ballot-scality

sayf-eddine-scality commented on Makefile:
```diff
@@ -17,6 +17,15 @@ ISO ?= $(BUILD_ROOT)/metalk8s.iso
 CALICO_CNI_PLUGIN_VERSION = 3.5.1
 CALICO_CNI_PLUGIN_BUILD = 1
 
+CONTAINER_IMAGES = \
```
should we put this list outside the makefile ? like we did for the calico plugin sources urls

alexandre-allard replied:
done

sayf-eddine-scality commented on scripts/get_container_images.sh:
```diff
@@ -0,0 +1,70 @@
+        execute_command \
+            "Saving container image $container_image to $image_archive" \
+            docker save "$container_image" \
+                --output "$ARCHIVE_DESTINATION/$image_archive"
```
Cant we compress the image ? ` | gzip -c > XX` , this will decrease the iso size and we can decompress in the load time

NicolasT replied:
Can we do better than `gz`? Should check whether `bz2` or `xz` are available on a bare-bones CentOS7 by default.

ballot-scality replied:
FYI, we have issue with bz with S3 archive because, for a very minimal gain, we take several more minutes just to uncompress it. @alexandre-allard-scality will now better about it, but I know he wanted to switch back to gzip for the docker images and the whole archive.

NicolasT replied:
Yeah. I experimented with both `bzip2` and `xz` last night. At least for compression, `bzip2` has a small gain compared to `gzip`, the gain from `xz` is more impactful (30+%). However, compression time with `xz` is a multiple (for some big factor) of the `gzip` compression time.

Maybe we should have some kind of 'devel' builds which use `gzip`, and `release` builds using `xz` one day.

alexandre-allard replied:
I've done some tests on gz, bz2 & xz for Federation stuff and it seems that gz is the best time to uncompress / size ratio.
But, if we don't care about the time used to uncompress, we should use xz which has the best compression in term of size.
That said, TS were asking for something faster, in general the overhead in the size of the archive is not really a problem, but the uncompression time can be painful.

NicolasT commented on Makefile:
```diff
@@ -17,6 +17,15 @@ ISO ?= $(BUILD_ROOT)/metalk8s.iso
 CALICO_CNI_PLUGIN_BUILD = 1
 
+CONTAINER_IMAGES = \
+	k8s.gcr.io/kube-apiserver@sha256:79b197b6a7334d2ad01ee82a9be82a39bfb50e33bc03263c58c4dfc4e81f71bc \
```
For all of these, please have a comment with them saying which tag this is. Also, if we pull-by-digest, do we get the tag in the metadata, i.e. when we import the image, does it show up with tag-name?

alexandre-allard replied:
Ok, will do.
No, there is no tag in the metadata when images are pulled by their digests.

NicolasT replied:
So, in the snippet I shared, I pull by tag, but export by digest, in an attempt to have the tags anyway (didn't check yet though *oops*)

NicolasT reviewed (changes requested):
As a general remark: not sure why there's a script in here, while instead retrieving images and saving them can be a simple set of `Makefile` rules?

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @ramanduh

* @alxf

* @tcarmet

* @slaperche-scality

* @cmonfort

* @zempashi

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @alexandre-allard-scality

* @ballot-scality

* @sayf-eddine-scality

* @lucieleonard

* @gdemonet

* @carlito-scality

* @teddyandrieux





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @ballot-scality

alexandre-allard committed:
Add pulling & saving of k8s container images
Fixes #563
Usage of target-specific variables assignment to factorize the build of images:
https://www.gnu.org/software/make/manual/html_node/Target_002dspecific.html

NicolasT commented on Makefile:
```diff
@@ -50,6 +50,14 @@ ALL = \
 	\
 	$(SCALITY_EL7_REPO) \
+	\
+	$(ISO_ROOT)/images/coredns.tar.gz \
```
Let's have the version/tag number in the tarball filename.

alexandre-allard replied:
done

NicolasT commented on Makefile:
```diff
@@ -232,3 +240,18 @@ $(SCALITY_EL7_REPODATA): $(SCALITY_EL7_RPMS) | $(PACKAGE_BUILD_CONTAINER)
+$(ISO_ROOT)/images/kube-scheduler.tar.gz: IMAGE=$(KUBE_SCHEDULER_IMAGE)
+$(ISO_ROOT)/images/nginx.tar.gz: IMAGE=$(NGINX_IMAGE)
+$(ISO_ROOT)/images/%.tar.gz:
+	mkdir -p $(@D)
```
Oh, new to me :)

alexandre-allard commented on Makefile:
```diff
@@ -6,7 +6,7 @@ MAKEFLAGS += -r
 SHELL := /bin/bash
 
-include VERSION
+include VERSION container_images.mk
```
Maybe one prefers 2 distinct includes ?

NicolasT replied:
#603 may change this. I think distinct `include` lines is a bit cleaner, but no blocker for me.

NicolasT reviewed (dismissed):
LGTM. We may revisit a couple of details later (e.g. making sure the hash we pull/save actually corresponds to the version we're expecting).

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @ramanduh

* @alxf

* @tcarmet

* @slaperche-scality

* @cmonfort

* @zempashi

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @alexandre-allard-scality

* @ballot-scality

* @sayf-eddine-scality

* @lucieleonard

* @gdemonet

* @carlito-scality

* @teddyandrieux





The following reviewers are expecting changes from the author, or must review again:

* @ballot-scality

alexandre-allard committed:
Tag images before saving them
Fixes #563

NicolasT commented on container_images.mk:
```diff
@@ -0,0 +1,27 @@
+COREDNS_IMAGE=k8s.gcr.io/coredns@sha256:3e2be1cec87aca0b74b7668bbe8c02964a95a402e45ceb51b2252629d608d03a
+COREDNS_IMAGE_NAME=coredns
+COREDNS_IMAGE_VERSION=1.1.3
```
It appears that `kubeadm` uses CoreDNS 1.2.2 when deploying K8s 1.11. Where did you get this version from?

NicolasT replied:
Huh no you're right, that's at least what it says in the ChangeLog. Odd.

NicolasT reviewed (dismissed):
:+1:

NicolasT commented:
@ballot-scality More comments on this one?

ballot-scality commented:
Didn't check yet.

Will do tomorrow morning

Le lun. 18 févr. 2019 à 20:58, Nicolas Trangez <notifications@github.com> a
écrit :

> @ballot-scality <https://github.com/ballot-scality> More comments on this
> one?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scality/metalk8s/pull/593#issuecomment-464858270>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AOjjwhTzZ6LEJbxv7dwrXcYks_LB7nNLks5vOwXegaJpZM4a9NpQ>
> .
>

-- 
 
<https://www.scality.com/download-idc-marketscape/?utm_source=ScalityEmail&utm_medium=Email&utm_campaign=2018Predictions>

ballot-scality commented on Makefile:
```diff
@@ -232,3 +240,38 @@ $(SCALITY_EL7_REPODATA): $(SCALITY_EL7_RPMS) | $(PACKAGE_BUILD_CONTAINER)
 		/entrypoint.sh buildrepo
+
+
+$(ISO_ROOT)/images/$(COREDNS_IMAGE_NAME)-$(COREDNS_IMAGE_VERSION).tar.gz: \
```
Overall I do agree with this block, it's just that having "=" without space makes it less "Makefile stylish".
No big deal, just saying.
For reference, I guess you used this : https://www.gnu.org/software/make/manual/html_node/Target_002dspecific.html ?
Given the knowledge about Makefile overall, would be cool to include it in the commit message IMO.

alexandre-allard replied:
Bad bash habits...
It's fixed

ballot-scality reviewed (dismissed):
LGTM, except the "style" of the Makefile variable assignment :)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @ramanduh

* @alxf

* @tcarmet

* @slaperche-scality

* @cmonfort

* @zempashi

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @alexandre-allard-scality

* @ballot-scality

* @sayf-eddine-scality

* @lucieleonard

* @gdemonet

* @carlito-scality

* @teddyandrieux

alexandre-allard commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
